### PR TITLE
recommend GHC2021

### PR DESCRIPTION
**Pull request by chris-martin**
Recommends using [GHC2021](https://downloads.haskell.org/ghc/9.2-latest/docs/html/users_guide/exts/control.html#extension-GHC2021) as the language version, and removes from our list extensions that are redundant to GHC2021.

Will wait a while on this to leave room for any discussion.

This description of the change is primarily based on the GHC 9.2 docs.

This implies the enabling of all the following extensions that the guide did not suggest previously:

- [`NamedFieldPuns`](https://downloads.haskell.org/ghc/9.2-latest/docs/html/users_guide/exts/record_puns.html?highlight=namedfieldpuns) - Allows contracting `A{x=x}` to `A{x}` in expressions and patterns
- [`NamedWildCards`](https://downloads.haskell.org/ghc/9.2-latest/docs/html/users_guide/exts/partial_type_signatures.html#extension-NamedWildCards) - Allow naming of wildcards (e.g. `_x`) in type signatures
- New numeric literal forms: `BinaryLiterals`, `HexFloatLiterals`, `NumericUnderscores`
- Allowing more explicit signatures: `InstanceSigs`, `KindSignatures`, `StandaloneKindSignatures`
- [`DeriveDataTypeable`](https://downloads.haskell.org/ghc/9.2-latest/docs/html/users_guide/exts/deriving_extra.html#extension-DeriveDataTypeable) - Allows stock deriving of the `Data` class
- [`EmptyDataDeriving`](https://downloads.haskell.org/ghc/9.2-latest/docs/html/users_guide/exts/empty_data_deriving.html#extension-EmptyDataDeriving) - Permits stock deriving for datatypes with no constructors
- Related to infix operators: [`PostfixOperators`](https://downloads.haskell.org/ghc/9.2-latest/docs/html/users_guide/exts/rebindable_syntax.html#extension-PostfixOperators), [`TupleSections`](https://downloads.haskell.org/ghc/9.2-latest/docs/html/users_guide/exts/tuple_sections.html#extension-TupleSections). [`TypeOperators`](https://downloads.haskell.org/ghc/9.2-latest/docs/html/users_guide/exts/type_operators.html#extension-TypeOperators)
- [`ImportQualifiedPost`](https://downloads.haskell.org/ghc/9.2-latest/docs/html/users_guide/exts/import_qualified_post.html#extension-ImportQualifiedPost) lets you write `import A qualified as B` instead of `import qualified A as B`
- Changes to type system: [`ConstraintKinds`](https://downloads.haskell.org/ghc/9.2-latest/docs/html/users_guide/exts/constraint_kind.html#extension-ConstraintKinds), [`ExistentialQuantification`](https://downloads.haskell.org/ghc/9.2-latest/docs/html/users_guide/exts/existential_quantification.html#extension-ExistentialQuantification), [`PolyKinds`](https://downloads.haskell.org/ghc/9.2-latest/docs/html/users_guide/exts/poly_kinds.html#extension-PolyKinds)
- [`GADTSyntax`](https://downloads.haskell.org/ghc/9.2-latest/docs/html/users_guide/exts/gadt_syntax.html#extension-GADTSyntax) - Adds an alternative syntax for defining datatypes
- [`EmptyCase`](https://downloads.haskell.org/ghc/9.2-latest/docs/html/users_guide/exts/empty_case.html#extension-EmptyCase) - Allows writing a `case` expression that doesn't have any cases (useful in weird circumstances to pattern match over something that's impossible e.g. `Void`)
- [`TypeSynonymInstances`](https://downloads.haskell.org/ghc/9.2-latest/docs/html/users_guide/exts/instances.html#extension-TypeSynonymInstances) - Allows types defined as `type` aliases to be referenced in the head of typeclass instance definitions
- `DoAndIfThenElse` - Actually I have no idea what this is, it's not documented in the manual.

GHC2021 also disables a few extensions that were previously enabled by Haskell2010:

- [`CUSKs`](https://downloads.haskell.org/ghc/9.2-latest/docs/html/users_guide/exts/poly_kinds.html#extension-CUSKs)
- [`DatatypeContexts`](https://downloads.haskell.org/ghc/9.2-latest/docs/html/users_guide/exts/datatype_contexts.html#extension-DatatypeContexts)

In the [latest](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/control.html#extension-GHC2021) GHC, there is an additional subtlety between Haskell2010 and GHC2021:

- GHC2021 does not enable [`DeepSubsumption`](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/rank_polymorphism.html#extension-DeepSubsumption).

### CODE REVIEW

**Comment by pbrisbin** [comment]
Added a few more reviewers.

Would it possible to list out what this does in the form of a difference against the current set? I'm interested if we're dropping anything we currently prescribe.

**Comment by chris-martin** [comment]
> list out what this does

Working on it in the PR description.

**Comment by pbrisbin** [comment]
> This implies the enabling of all the following extensions that the guide did not suggest previously

Sorry I didn't fully read this. Are you saying here that `GHC2021` is a strict _superset_ of the current guide and only adds things as you've described, taking nothing away?

**Comment by chris-martin** [comment]
> GHC2021 is a strict superset

There are a few removals that I've missed, updating the list now.

**Comment by chris-martin** [comment]
Updated the list to reflect only what's in the GHC 9.2 documentation.

**Comment by chris-martin** [comment]
In the [latest](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/control.html#extension-GHC2021) docs, there are a few additional subtleties between Haskell2010 and GHC2021.

**Review by mjgpy3 — APPROVED** [review]
I'm for it, embrace the future!

File: haskell-style.md [hunk 1]
```diff
@@ -234,35 +234,23 @@ Sort your exports, unless the order matters in your desired Haddock output.
 
 **Automated**: sorting and formatting of extension pragmas.
 
-- All Haskell packages MUST use the following `default-extensions`:
+- All Haskell packages MUST use the following language version and `default-extensions`:
```
Comment (pbrisbin) [inline haskell-style.md]: Thoughts on:

```suggestion
- All Haskell packages MUST use the following `language` and `default-extensions`:
```

?

If we're going to speak of one of them as `code`, might as well do them both that way.

**Review by pbrisbin — APPROVED** [review]

**Review by z0isch — APPROVED** [review]

**Review by danroyo — APPROVED** [review]

**Review by stackptr — APPROVED** [review]

**Comment by joelmccracken** [comment]
https://gitlab.haskell.org/haskell/prime/-/wikis/do-and-if-then-else/diff?version_id=c90e3fcd2c434ae6dd8704026d4244bd66015be1 seems like really very little docs

**Pull request merged**